### PR TITLE
136287611 - Allowed file types

### PR DESCRIPTION
**Pull request by nisevi**
PT Story: https://www.pivotaltracker.com/story/show/136287611

Changes proposed in this pull request:
1. Mime types allowed on 'UploadedFile' model are picked from constant;
2. Same constant used on controller to let the view know which file extensions are allowed;

Screenshots:
image

### CODE REVIEW

**Review by thesuss — APPROVED** [review]

File: app/views/shf_applications/_shf_application_fields.html.haml [hunk 1]
```diff
@@ -46,11 +46,15 @@
...
       %h3.upload-button-pre= t('shf_applications.new.upload_files')
 
+
       %span{class: "btn btn-warning btn-file"}
         = t('shf_applications.new.upload_button')
         = f.file_field 'actual_files[]', multiple: true, name: 'uploaded_file[actual_files][]', class: 'upload-field', id: 'upload-button'
 
       - submit_label = @shf_application.new_record? ? t('submit') : t('save')
+      %p= t('shf_applications.new.upload_allowed_file_types')
+      %p= @allowed_file_types
```
[the remark this answers is not in the document]
Comment (patmbolger) [reply app/views/shf_applications/_shf_application_fields.html.haml]: Please add an `accept` option to the `file_field` (line 51) so that the user can only select valid file types in the upload-file dialog box.

If you find a nice way to test this added item, it would be great.  But, that is not critical for this.

**Review by patmbolger — CHANGES_REQUESTED** [review]

File: app/views/shf_applications/_shf_application_fields.html.haml (hunk 1 above)
[the remark this answers is not in the document]
Comment (patmbolger) [reply app/views/shf_applications/_shf_application_fields.html.haml]: BTW, adding the above item would also address https://www.pivotaltracker.com/story/show/153360586.

**Comment by nisevi** [comment]
Allowed file type extensions are being displayed to the user:

image

Only display to the user the allowed files that can be uploaded:

image

This are all the files on that folder:

image

File: app/views/shf_applications/_shf_application_fields.html.haml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nisevi) [reply app/views/shf_applications/_shf_application_fields.html.haml]: @patmbolger the `accept` option was added to the `file_field` for the user to only see the allowed file types.

**Comment by nisevi** [comment]
`Cucumber` is failing on `SemaphoreCI` because I think that is trying to find a type of file that we are not displaying any more, since we are specifying which file types the user should see as default. I mean the user always can select `All files` and then it will see other file types to upload, so I'm working on the cucumber tests now :+1:

**Comment by nisevi** [comment]
These are the failing scenarios:

```
Failing Scenarios:
cucumber features/create_membership_application.feature:133 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#1)
cucumber features/create_membership_application.feature:134 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#2)
cucumber features/create_membership_application.feature:135 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#3)
cucumber features/create_membership_application.feature:136 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#4)
cucumber features/create_membership_application.feature:137 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#5)
cucumber features/create_membership_application.feature:138 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#6)
cucumber features/create_membership_application.feature:153 # Scenario Outline: Apply for membership: company number wrong length, Scenarios (#1)
cucumber features/create_membership_application.feature:156 # Scenario: Cannot change locale if there are errors in the new application
cucumber features/edit_shf_application.feature:35 # Scenario: Applicant makes mistake when editing his own application
cucumber features/edit_shf_application.feature:90 # Scenario: Cannot change locale if there are errors in the application
cucumber features/emails/new-application-email-to-admins.feature:74 # Scenario: User submits a new application app with bad info so it is not created, so no email sent [SAD PATH]
cucumber features/emails/new-application-gets-ack-email.feature:38 # Scenario: User submits a new application app with bad info so it is not created, so no email sent [SAD PATH]
cucumber features/size_validation.feature:18 # Scenario: New application - Uploads a file that is too large
cucumber features/size_validation.feature:42 # Scenario: Existing application - Uploads a file that is too large
cucumber features/size_validation.feature:60 # Scenario: Size error message in English when locale = en
cucumber features/size_validation.feature:71 # Scenario: Size error message in sveriges when locale = sv
cucumber features/size_validation.feature:80 # Scenario: Switching locales - Size error message in the right language
cucumber features/upload.feature:92 # Scenario: Try to upload a file with unacceptable content type
cucumber features/view_all_companies.feature:143 # Scenario: Visitor sees multiple regions and kommuns for a company in the companies list
cucumber features/view_all_companies.feature:209 # Scenario: Pagination: Set number of items per page

```

**Comment by nisevi** [comment]
The problem wasn't related to my previous assumption, the real problem was that on `update` and `create` actions, the object holding the allowed types didn't exist, and when the file type was wrong the view was trying to use it.

File: app/controllers/shf_applications_controller.rb [hunk 2]
```diff
@@ -218,6 +218,7 @@ def new_file_uploaded(params)
                                            filename: @uploaded_file.actual_file_file_name))
           successful = successful & true
         else
+          @shf_application.uploaded_files.delete(@uploaded_file)
```
Comment (nisevi) [inline app/controllers/shf_applications_controller.rb:221]: @patmbolger I deleted the object because even if the `@uploaded_file` was invalid (here we are in the `else` statement for the `@uploaded_file.valid?` validation) it was hanging there as part of the uploaded files when we were back at the view. Please let me know if you consider that this change should be done in another PR.

Comment (patmbolger) [reply app/controllers/shf_applications_controller.rb:221]: No problem - that's fine doing it here.

**Comment by patmbolger** [comment]
These changes look good.

In addition, please use the list of allowed file types (from the model) in the rspec validation test.

Right now these are repeated in the test:
```
it { should validate_attachment_content_type(:actual_file)
                    .allowing('image/jpeg', 'image/gif', 'image/png',
                              'text/plain',
                              'text/rtf',
                              'application/pdf',
                              'application/msword',
                              'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                              'application/vnd.ms-word.document.macroEnabled.12')
                    .rejecting('bin', 'exe') }
```

**Review by patmbolger — CHANGES_REQUESTED** [review]
Please see comment re change to the rspec test.

**Comment by nisevi** [comment]
@patmbolger I updated the rspec test for using the use the list of allowed file types from the model :+1:

**Review by patmbolger — APPROVED** [review]
